### PR TITLE
Resource state retry

### PR DESCRIPTION
**Pull request by gaganapplatix**
Fix for https://github.com/argoproj/argo/issues/627

### CODE REVIEW

File: workflow/executor/resource.go [hunk 1]
```diff
@@ -80,51 +92,151 @@ func (we *WorkflowExecutor) WaitResource(resourceName string) error {
...
+				// always block and we need to call that to get the exit code of the process. So we will unconditionally
+				// call exec.Process.Kill and then assume that wait will not block after that. Two things may happen:
+				// 1. Process already exited and kill does nothing (returns error which we ignore) and then we call
+				//    Wait and get the proper return value
+				// 2. Process is running gets, killed with exec.Process.Kill call and Wait returns an error code and we give up
+				//    and don't retry
+				_ = cmd.Process.Kill()
+
+				msg = fmt.Sprintf("Command for kubectl get -w for %s exited. Getting return value using Wait", resourceName)
+				log.Warnf(msg)
```
Comment (jessesuen) [inline workflow/executor/resource.go]: Nit: the pattern of putting msg into a variable and passing msg to log.XXXXf is a little hard to parse. I think it would be fine if we were reusing the message (e.g. in a returned error), but if all we are doing is logging the message, can we just do:
```
log.Warnf("Command for kubectl get -w for %s exited. Getting return value using Wait", resourceName)
```
and save a bunch of lines? Also for code base searches it will be easier to grep for what we are warning/erroring about.

File: workflow/executor/resource.go [hunk 2]
```diff
@@ -80,51 +92,151 @@ func (we *WorkflowExecutor) WaitResource(resourceName string) error {
 		failReqs, _ = failSelector.Requirements()
 	}
 
-	cmd := exec.Command("kubectl", "get", resourceName, "-w", "-o", "json")
-	stdout, err := cmd.StdoutPipe()
+	ctrlChannel := make(chan string)
+	resultChannel := make(chan waitResult)
+	go checkResourceState(ctrlChannel, resultChannel, resourceName, successReqs, failReqs)
+
+	err := wait.ExponentialBackoff(wait.Backoff{Duration: (time.Second * 5), Factor: 4.0, Steps: 5},
```
[the remark this answers is not in the document]
Comment (gaganapplatix) [reply workflow/executor/resource.go]: So the first iteration in the Exponential Backoff is with 0 delay so 5 steps are 0, 5, 20, 80, 320 so max should be a little over 5 min.

File: workflow/executor/resource.go [hunk 3]
```diff
@@ -80,51 +92,151 @@ func (we *WorkflowExecutor) WaitResource(resourceName string) error {
...
+			if result.result == retryStr {
+				log.Infof("Waiting for resource %s resulted in retryable error %v", resourceName, result.err)
+				return false, nil
+			}
+
+			log.Warnf("Waiting for resource %s resulted in non-retryable error %v", resourceName, result.err)
+			return false, result.err
+		})
+
+	close(ctrlChannel)
```
[the remark this answers is not in the document]
Comment (gaganapplatix) [reply workflow/executor/resource.go]: This forces the go routine to stop executing and exit. Since after ExponentialBackoff returns we are done with all the iterations that we want to run. So when eventually the go routine goes back to waiting for the ctrlChannel it will see the close channel and exit itself.

File: workflow/executor/resource.go [hunk 4]
```diff
@@ -80,51 +92,151 @@ func (we *WorkflowExecutor) WaitResource(resourceName string) error {
...
+		}
 	}
-	defer func() {
-		_ = cmd.Process.Kill()
-	}()
+
+	return err
+}
+
+func checkResourceState(ctrlChannel <-chan string, resultChannel chan<- waitResult, resourceName string,
```
[the remark this answers is not in the document]
Comment (gaganapplatix) [reply workflow/executor/resource.go]: This method is the go routine that does the kubectl command start and then waits for the json. If it encounters errors or success or failure it communicates to the anonymous function in ExponentialBackoff on the resultChannel which then interprets the errors for ExponentialBackoff.

File: workflow/executor/resource.go (hunk 1 above)
Comment (gaganapplatix) [reply workflow/executor/resource.go]: Fixed

File: workflow/executor/resource.go [hunk 5]
```diff
@@ -80,51 +92,151 @@ func (we *WorkflowExecutor) WaitResource(resourceName string) error {
...
+			result := <-resultChannel
+
+			if result.result == "" {
+				msg := fmt.Sprintf("Waiting for resource %s resulted in channel close", resourceName)
+				log.Warnf(msg)
+				return false, errors.InternalError(msg)
+			}
+
+			if result.err == nil {
+				log.Infof("Returning from successfull wait for resource %s", resourceName)
```
[the remark this answers is not in the document]
Comment (gaganapplatix) [reply workflow/executor/resource.go]: Fixed

**Comment by jessesuen** [comment]
I know I was the one to suggest channels... but it's hard to understand the use of the backwards resultChannel to communicate back to the wait.ExponentialBackoff. I think perhaps this should all be done in a synchronous function that blocks and returns success/failure/and whether or not to retry again. I think it will make this easier to understand.

**Comment by jessesuen** [comment]
This would also negate the need for the start channel since that would be handled externally by the wait logic

**Comment by gaganapplatix** [comment]
Yeah I was thinking the same thing, the channel stuff is pointless since we are synchronously running everything single threaded, so I have changed it to be just inline function call from ExponentialBackoff and only single loop for json reading. Seems to work fine
